nemo/core/classes/common.py

```diff
@@ -74,11 +75,39 @@ def _attach_and_validate_output_types(self, out_objects):
         # TODO: Properly implement this
         if self.output_types is not None:
             out_types_list = list(self.output_types.items())
+
+            # First convert all outputs to list/tuple format to check correct number of outputs
+            if type(out_objects) in (list, tuple):
+                out_container = out_objects
+            else:
+                out_container = [out_objects]
+
+            if len(self.output_types) != len(out_container):
+                raise TypeError(
+                    "Number of output arguments provided ({}) is not as expected ({})".format(
+                        len(out_container), len(self.output_types)
+                    )
+                )
+
+            # Attach types recursively, if possible
             if not isinstance(out_objects, tuple) and not isinstance(out_objects, list):
-                out_objects.neural_type = out_types_list[0][1]
+                try:
+                    out_objects.neural_type = out_types_list[0][1]
+                except Exception:
```

Comment:
Let's not catch Exception, can we either find out exactly what error is raised and catch that? Alternatively, can we somehow define our own exception?

Reply:
It depends on the type of object you submit, and how it reacts to an attempt to assign a object instance to it. List behaves differently than tuple, than base data types, than structs, than data classes. There's too many combinations to check explicitly.

Reply:
As to our own exception, we first need to catch something in order to raise it with our own. These exceptions comes from python core itself, no chance to override or wrap